### ISSUE TITLE
Prisma bundling broken after release v3.3.16

### ISSUE THREAD

**Issue by Liamandrew** [anomalyco/sst#5077]
I have a similar setup to the [Prisma lambda example ](https://github.com/sst/sst/tree/dev/examples/aws-prisma-lambda), however after release [v3.3.16](https://github.com/sst/sst/releases/tag/v3.3.16) I've discovered that the bundling no longer works as before and I receive the following error response:

```
Prisma Client could not locate the Query Engine for runtime \"rhel-openssl-3.0.x\".\n\nThis is likely caused by a bundler that has not copied \"libquery_engine-rhel-openssl-3.0.x.so.node\" next to the resulting bundle.\nEnsure that \"libquery_engine-rhel-openssl-3.0.x.so.node\" has been copied next to the bundle or in \"../../node_modules/.prisma/client\".
```

I've tried a few different configuration options but so far my only workaround is to downgrade back to v3.3.14.

**Comment by jayair** [comment]
When do you get this error?

**Comment by Liamandrew** [comment]
Any time a prisma query is executed during runtime

**Comment by jayair** [comment]
Are you Prisma versions the same as the example? https://github.com/sst/sst/blob/dev/examples/aws-prisma-lambda/package.json

**Comment by Liamandrew** [comment]
@jayair I'm using the ~latest~ Prisma version `5.22.0`.

I'm also using a monorepo in case that's relevant.

**Comment by Liamandrew** [comment]
I've also tried the latest prisma version `6.0.1` but still get the same error that the query engine could not be found during runtime.

**Comment by jayair** [comment]
Might need a repo to reproduce then, maybe it's related to the monorepo setup.

**Comment by Liamandrew** [comment]
@jayair https://github.com/Liamandrew/sst-prisma-example

This is a minimal reproducible example of my setup. A turborepo with prisma as a package in `packages/database`.

**Comment by jayair** [comment]
I'll take a look.

**Comment by jayair** [comment]
I haven't had a chance to try your repo yet. But I retried the example we have with the latest SST and it works.

For your case I would check if this `copyFiles: [{ from: "node_modules/.prisma/client/" }]` is generated after you run `prisma generate`. And the `from` path is relative to the location of the `sst.config.ts`.

**Comment by Liamandrew** [comment]
@jayair I've updated to the latest versions and it is still broken. 

I think we agreed that the example isn't the same because it is not a monorepo.

I'd appreciate it if you could actually take a look since I've put together the [steps required to deploy and reproduce the issue](https://github.com/Liamandrew/sst-prisma-example?tab=readme-ov-file#sst), and I've pinpointed the exact version where it broke.

**Comment by jayair** [comment]
I'll have to find some time to do this. In the meantime can you check:

> copyFiles: [{ from: "node_modules/.prisma/client/" }] is generated after you run prisma generate

**Comment by cmaycumber** [comment]
I'm running into this on a monorepo as well. 

I have the following implemented:

```ts
copyFiles: [{ from: "node_modules/.prisma/client/" }]
```

**Comment by cmaycumber** [comment]
Ripping the db folder outside of our `apps/backend` package and bringing it top level, similar to how the prisma example is structured, solves the issue oddly enough.

**Comment by Liamandrew** [comment]
> I'll have to find some time to do this. In the meantime can you check:
> 
> > copyFiles: [{ from: "node_modules/.prisma/client/" }] is generated after you run prisma generate

Thanks.

I'm not exactly sure where/what to check, but digging around the `.sst` directory, I can see that a path `node_modules/.prisma` is generated in `.sst/artifacts/MyApi-src`, however it is empty

**Comment by cmaycumber** [comment]
Running into the above ^ myself. Only seeing the error when deployed. I thought the issue is it only works for single files. 

Even in the SST docs:

```md
Add additional files to copy into the function package. Takes a list of objects with from and to paths. These will be copied over before the function package is zipped up.

Copying over a single file from the src directory to the src/ directory of the function package.
```

I think the docs and/or the example need to be updated to reflect this if it's true. I tried creating all the file paths with globby but was still unable to get the folder to populate for some reason.

**Comment by cmaycumber** [comment]
It seems like there's a bug in copying files in general. 

Here's my $transform function:
```js
// Apply default copy all the files to the layer...
    $transform(sst.aws.Function, (args, opts) => {
      args.link ??= secrets;

      args.copyFiles ??= [
        {
          from: 'db/schema.prisma',
          to: 'schema.prisma',
        },
        ...prismaFiles.map((file) => ({
          from: file,
          to: file,
        })),
      ];

      args.nodejs ??= {
        sourcemap: isDeployment,
        // Crawlee cannot be bundled so we need to install it here
        install: ['crawlee', 'playwright'],
        esbuild: {
          keepNames: false,
          external: ['crawlee'],
        },
      };

      args.environment = {
        // Add this so that AWS will re-use TCP connections instead of re-connecting every time
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
        // Disable X-Ray warnings for missing context (happens because we declare clients outside of handler)
        AWS_XRAY_CONTEXT_MISSING: 'IGNORE_ERROR',
        // Store the GITHUB_SHA in the environment so it can be referenced as our release version for sentry (set by CI)
        GITHUB_SHA: process.env.GITHUB_SHA ?? '',
        // Enable source maps for sentry
        NODE_OPTIONS: isDeployment ? '--enable-source-maps' : '',
        // Whether we are running locally
        ...($dev ? { IS_LOCAL: 'true' } : {}),
      };
    });
```

I added the file to test and it didn't seem to be working:
```
        {
          from: 'db/schema.prisma',
          to: 'schema.prisma',
        },   
```

**Comment by jayair** [comment]
@cmaycumber thanks for digging into this.

Can you give me some steps to reproduce this copy file issue?

**Comment by cmaycumber** [comment]
> @cmaycumber thanks for digging into this.
> 
> Can you give me some steps to reproduce this copy file issue?

Yep I’ll do my best. Here’s how I’m going about checking for this.

Folder structure:
- apps/frontend
- apps/backend
- Infra
- packages
- db

Steps:
1. Create a function using SST
2. Leverage transform to copy files (not working)
3. Check the .sst/artifacts function src folder
4. Try individual function copy files (not working)
5. Repeat step 3

I hope this helps. I’ll try a fresh project as well with similar settings to see if it compares.

**Comment by cmaycumber** [comment]
Here's a very minimal reproduction: 

https://github.com/cmaycumber/prisma-copy-files-bug

I think something is removing the files after being added but this is pretty close to the prisma example I'm not sure how that one works.

**Comment by cmaycumber** [comment]
Temporarily solved this with a lambda layer: 

```js
import * as aws from '@pulumi/aws';
import * as pulumi from '@pulumi/pulumi';
import * as child_process from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

// Create temporary directories for bundling
const prismaLayerDir = path.join(__dirname, 'prisma-layer');
const crawleeLayerDir = path.join(__dirname, 'crawlee-layer');

// Clean up existing layer directories if they exist
if (fs.existsSync(prismaLayerDir)) {
  fs.rmSync(prismaLayerDir, { recursive: true });
}
if (fs.existsSync(crawleeLayerDir)) {
  fs.rmSync(crawleeLayerDir, { recursive: true });
}

// Set up prisma layer
const prismaNodeModulesDir = path.join(prismaLayerDir, 'nodejs', 'node_modules');
fs.mkdirSync(prismaNodeModulesDir, { recursive: true });

// Copy Prisma client files
const prismaClientPath = path.join(__dirname, '../..', 'node_modules/.prisma/client');
fs.cpSync(prismaClientPath, path.join(prismaNodeModulesDir, '.prisma/client'), { recursive: true });

// Create prisma layer
export const prismaLayer = new aws.lambda.LayerVersion('prisma-layer', {
  compatibleRuntimes: ['nodejs18.x', 'nodejs20.x'],
  code: new pulumi.asset.AssetArchive({
    '.': new pulumi.asset.FileArchive(prismaLayerDir),
  }),
  description: 'Lambda layer containing Prisma client',
  layerName: 'prisma-layer',
});
```

and applying the engine path: 
```js
// Prisma query engine library
        ...(!$dev
          ? {
              PRISMA_QUERY_ENGINE_LIBRARY:
                '/opt/nodejs/node_modules/.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
            }
          : {}),
```

**Comment by jayair** [comment]
How do I try your repo @cmaycumber? 

```
npm install
npm run deploy
```

The deploy script is not running sst?

```
    "build": "esbuild --bundle --outfile=./dist/index.js --platform=node --target=node20 ./src/index.ts",
    "deploy": "run-s build zip update",
    "update": "aws lambda update-function-code --zip-file fileb://lambda.zip --function-name hello",
    "zip": "zip -j lambda.zip dist/index.js",
```

**Comment by cmaycumber** [comment]
Oh sorry. I was using `bunx sst deploy`. I scraped this together in a hurry.

**Comment by cmaycumber** [comment]
Those scripts don't do anything

**Comment by jayair** [comment]
I'll try it out.

**Comment by jayair** [comment]
Saw this in Discord, passing this along https://discord.com/channels/983865673656705025/1331969834103078954/1332349076154351636

**Comment by Liamandrew** [comment]
It's a bit hard to follow what the exact solution is in that thread, but it looks like that person doesn't have a typical prisma setup (i.e. they are outputting their generated client in a custom directory) so it most likely won't apply to all.

I've noticed that if the `to` path has a `.` prefix in any of the subpaths then it will not include that file(s) correctly.

For example, this doesn't include the generated client in the assets because of `.prisma`

```
{
  from: 'node_modules/.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
  to: 'node_modules/.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
},
```

This will include the generated client in the assets:
```
{
  from: 'node_modules/.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
  to: 'node_modules/prisma/libquery_engine-rhel-openssl-3.0.x.so.node',
}
```

But the generated client needs to be in either `node_modules/.prisma/client` or `.prisma/client` so this _might_ be the problem.

If I downgrade to sst version 3.3.14 like I originally mentioned, this isn't a problem and the following configuration is all I need for it work:

```
{
  from: 'node_modules/.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
  to: '.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
},
``

**Comment by jayair** [comment]
Interesting, sadly I haven't had a chance to look at it yet. But this will help.

**Comment by Liamandrew** [comment]
Okay, some more information after I've dedicated some time to get a workaround in place because of FOMO not being able to upgrade SST for so long.

Using the knowledge from the discord thread and the findings that a path with `.prisma` doesn't work for now, this is how I've worked around it:

Added a custom client output path to `schema.prisma` so that the default path `.prisma` is not used:

```
output        = "generated"
```

Then in my ApiGatewayV2 route handler I've added the following configuration to copy the generated client engine:

```
copyFiles: {
  from: 'packages/database/prisma/generated/libquery_engine-rhel-openssl-3.0.x.so.node',
  to: 'prisma/generated/libquery_engine-rhel-openssl-3.0.x.so.node',
}
```

And it works 🎉

**Comment by corybill** [comment]
I have same issues listed here. I am not in a mono repo though. I am using remix though.

I cannot use the above workaround because there deoesn't appear to be a "copyFiles" for Remix even though it is using a Lambda.

I am more or less doing the serverless version here: https://sst.dev/docs/start/aws/remix with Prisma, follwing this: https://sst.dev/docs/examples#prisma-in-lambda.

But I cannot do the copyFiles step.

Is there a way to use copyFiles with https://sst.dev/docs/component/aws/remix. It is deploying to serverless so should it have Function params exposed?

Here is a simple example that shows the error.
https://github.com/corybill/sst-remix-prisma-broken

Thanks!!

**Comment by Liamandrew** [comment]
@corybill I haven't used remix, but looking at the docs the API seems to be similar to nextjs. In which case, you should be able to use this to configure copyFiles with the server function:
https://sst.dev/docs/component/aws/remix#transform-server

**Comment by Liamandrew** [comment]
I feel like I'm so close to a proper fix without needing to play around with a custom prisma output, I just need a final push to get over the line.

If I start from the beginning, the problem occurred after this [commit](https://github.com/sst/sst/commit/b10b3a08baa5c601a959d926ae494200977fe6a3) which changed the bundle path to be at the root of the outputted build asset instead of being nested. This means that for monorepos (for example, in my case), the change caused this structure:

`apps/api/src/bundle.mjs` to `bundle.mjs`

Forget what I said about the prisma file not being included if the directory started with `.prisma` [here](https://github.com/sst/sst/issues/5077#issuecomment-2617216351), I believe that was incorrect and I simply missed that fact that the directory was hidden from file viewers.

However, it does seem that for monorepos, it is important that the outputted folder structure is similar to the monorepo setup for prisma, otherwise it struggles to find the relative location of the prisma client (for some reason).

I could also see that it worked fine for the nextjs build but not the regular function build. Taking a look at the source code for [opennextjs-aws](https://github.com/opennextjs/opennextjs-aws/blob/068ce66f31964543c88c7bdd1fa53f7adb76ab3d/packages/open-next/src/build/createServerBundle.ts#L242), there is some special handling during the build for monorepos that preserves this entry point and directory structure, leading it to work fine.

This makes me believe that I just need to provide some parameters to the [esbuild](https://sst.dev/docs/component/aws/function/#nodejs-esbuild) property of the function to preserve the directory structure so that the default pattern returns to the previous pattern before the above commit:

 `bundle.mjs` back to `apps/api/src/bundle.mjs`

I've tried playing around with `outfile` & `outdir` but they either don't do anything or aren't being applied. This is where I'm stuck...

**Comment by corybill** [comment]
@Liamandrew  My example now works thanks to you telling me where to put copyFiles command. Much appreciated!! I am following this for the Prisma setup. It calls out the need for the copyFiles command and gives an example on how to do it.

https://sst.dev/docs/examples#prisma-in-lambda

// NO LONGER BROKEN
https://github.com/corybill/sst-remix-prisma-broken

I'll leave my example available in case anybody else needs a reference.

One more note, I have seen 1 or 2 SST deploys since yesterday. Not sure if the updated versions are fixing something.

**Comment by Liamandrew** [comment]
Another update, an even better result now - no need to play around with esbuild properties. I have managed to make Prisma work in a monorepo with [SST Function ](https://sst.dev/docs/component/aws/function/) by using the environment variable [PRISMA_QUERY_ENGINE_LIBRARY](https://www.prisma.io/docs/orm/reference/environment-variables-reference#prisma_query_engine_library). So my config looks like this:

```
{
  copyFiles: [
    {
       from: 'node_modules/.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
       to: '.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
    },
  ],
  environment: {
    PRISMA_QUERY_ENGINE_LIBRARY:
      '.prisma/client/libquery_engine-rhel-openssl-3.0.x.so.node',
}
```

Just to re-iterate, from my findings this is only necessary when using **Prisma + SST Function in a monorepo** since it seems that the build of a function doesn't consider the monorepo folder structure and generates the output in the root of the outdir which happens to break the prisma client lookup.

The reason [nextjs](https://github.com/opennextjs/opennextjs-aws/blob/068ce66f31964543c88c7bdd1fa53f7adb76ab3d/packages/open-next/src/build/createServerBundle.ts#L242) works is because internally they handle the monorepo path structure which incidentally appears to maintain the prisma lookup logic.

Anyway, I am quite relieved now as I think this is a much cleaner workaround to get it working than playing around with custom prisma output paths. Hopefully this saves some time and headaches for others.
